byterun/minor_gc.c

```diff
@@ -434,25 +487,35 @@ CAMLexport value caml_promote(struct domain* domain, value root)
     caml_scan_stack (forward_pointer, st.promote_domain, domain_state->current_stack);
 
     /* Scan major to young pointers. */
-    for (r = remembered_set->major_ref.base; r < remembered_set->major_ref.ptr; r++) {
+    for (r = remembered_set->major_ref.base;
+         r < remembered_set->major_ref.ptr; r++) {
       value old_p = **r;
-      if (Is_block(old_p) && young_ptr <= old_p && old_p < young_end) {
+      if (Is_block(old_p) && is_in_interval(old_p,young_ptr,young_end)) {
         value new_p = old_p;
         forward_pointer (st.promote_domain, new_p, &new_p);
         if (old_p != new_p)
           caml_atomic_cas_raw (*r,old_p,new_p);
       }
     }
 
+    /* Scan ephemeron ref table */
+    for (re = remembered_set->ephe_ref.base;
+         re < remembered_set->ephe_ref.ptr; re++) {
+      value* key = &Op_val(re->ephe)[re->offset];
```

Comment:
This might not be a key: what if `re->offset == CAML_EPHE_DATA_OFFSET` ?

Reply:
In this case, we are doing a promotion. In particular, nothing is GCed. We handle ephemeron data just like we'd handle keys -- forward the reference in `ephe_ref` table if the object happens to be promoted.

Reply:
OK, I see. The variable name `key` was a red flag to me, but I agree this looks correct.